### PR TITLE
Add literal type to union literal

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
https://github.com/flyteorg/flyte/issues/2817

```python=
class Color(Enum):
    RED = "red"
    GREEN = "green"
    BLUE = "blue"


@task
def coalesce(a:  Union[List[str], str, Color]) -> str:
    if type(a) == Color:
        print(a.name)
        return a.name
    print(a)
    return a


@workflow
def wf(a:  Union[List[str], str, Color]) -> str:
    coalesced = coalesce(a=a)
    return coalesced


if __name__ == "__main__":
    wf(a=Color("red"))
```
Take the above code as an example. If we run `pyflyte run --remote example_union.py wf --a "flyte"`, we'll get the below error.

```bash
Request failed with status code 400 invalid a input wrong type, expected union_type:<variants:<collection_type:
<simple:STRING > structure:<tag:"Typed List" > > variants:<simple:STRING structure:<tag:"str" > > variants:<enum_type:
<values:"red" values:"green" values:"blue" > structure:<tag:"DefaultEnumTransformer" > > > , got simple:STRING instead
```
That's because literal(string) can be converted to str or enum, which is [ambiguous](https://github.com/flyteorg/flytepropeller/blob/0301f2a216001bdb988eef4ae58293f525076ad3/pkg/compiler/validators/typing.go#L214-L222). To address this, we should add [literal type](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/core/literals.proto#L58) in Union type, and it can force [validator](https://github.com/flyteorg/flytepropeller/blob/0301f2a216001bdb988eef4ae58293f525076ad3/pkg/compiler/validators/typing.go#L33-L35) to cast "flyte" to string 

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Comment by eapolinario** [comment]
Can you add a unit test?

**Review by eapolinario — DISMISSED** [review]

**Comment by wild-endeavor** [comment]
this is dependent on the propeller pr going out first right?

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
